sparse/core.py

```diff
@@ -1301,19 +1301,22 @@ def concatenate(arrays, axis=0):
     assert all(x.shape[ax] == arrays[0].shape[ax]
                for x in arrays
                for ax in set(range(arrays[0].ndim)) - {axis})
+    nnz = 0
+    dim = sum(x.shape[axis] for x in arrays)
+    shape = list(arrays[0].shape)
+    shape[axis] = dim
+
+    coords_dtype = np.min_scalar_type(max(shape) - 1) if len(shape) != 0 else np.uint8
     data = np.concatenate([x.data for x in arrays])
-    coords = np.concatenate([x.coords for x in arrays], axis=1)
+    coords = np.concatenate([x.coords for x in arrays], axis=1).astype(coords_dtype)
```

Comment:
Is it better to call `.astype(...)` before concatenation on all of `arrays` or after as is done here?

Reply:
I would think after. Because it would be applied to a bigger array and thus the overhead would be small. Memory would be better, too, as we are just passing already-allocated arrays to `concatenate`, otherwise we'd have to allocate for each temp array.

Incidentally, `astype` just returns the original array if it is already the correct dtype.